pyfar/dsp/dsp.py

```diff
@@ -2669,3 +2669,168 @@ def normalize(signal, reference_method='max', domain='auto',
+        above). In case of a linear correlation, the time lags are in the
+        interval ``[-signal_2.n_samples + 1, signal_1.n_samples - 1]``. In case
+        of the cyclic correlation, they are in the interval
+        ``[-(signal_1.n_samples // 2) + 1, signal_1.n_samples // 2]`` if the
+        signals have an even number of samples, and in the interval
+        ``[-(signal_1.n_samples // 2), signal_1.n_samples // 2]`` if the
+        signals have an odd number of samples.
+    argmax : numpy array
+        The delays applied to `signal_2` which maximize the correlation between
+        the signals, i.e., :math:`\arg \max_{l} \, |c[l]|` where the absolute
+        :math:`|\cdot|` is required for complex-valued time signals. The shape
+        of `argmax` corresponds to the `cshape` to which the input signals were
+        broadcaseted.
+
+    Examples
+    --------
+
+    Linear and cyclic auto-correlation of a perfect sequence. Perfect sequences
+    have unit auto-correlation for :math:`l=0` and zero auto-correlation
+    otherwise
+
+    .. plot::
+
+        >>> import pyfar as pf
+        >>>
+        >>> signal = pf.signals.linear_perfect_sweep(2**8)
+        >>>
+        >>> for mode in ['linear', 'cyclic']:
+        >>>     cor, lags, _ = pf.dsp.correlate(signal, signal, mode)
+        >>>     result = pf.TimeData(cor / pf.dsp.energy(signal), lags)
+        >>>     ax = pf.plot.time(result, label=mode)
+        >>>
+        >>> ax.set_xlabel('time lag in samples')
+        >>> ax.set_ylabel('auto correlation')
+        >>> ax.legend()
+
+    Compute the lags (delay) that are required to maximize the
+    cross-correlation between a one and multi-dimensional signal
+
+    .. plot::
+
+        >>> import pyfar as pf
+        >>> import numpy as np
+        >>>
+        >>> # one-dimensional signal: impulse with zero delay
+        >>> signal_1 = pf.signals.impulse(5, 0)
+        >>>
+        >>> # multi-dimension signal with cshape = (2, 2):
+        >>> # impulses with non-zero delays
+        >>> delays = np.array([[0, 1], [2, 3]], dtype=int)
+        >>> signal_2 = pf.signals.impulse(5, delays)
+        >>>
+        >>> _, _, argmax = pf.dsp.correlate(signal_1, signal_2, 'linear')
+
+    In this case the lags correspond to the negative delays:
+    ``argmax = [[0, -1], [-2, -3]]``.
+    """
+
+    # check input
+    if type(signal_1) is not pyfar.Signal or \
+            type(signal_2) is not pyfar.Signal:
+        raise TypeError("signal_1 and signal_2 must be pyfar.Signal objects")
+
+    if signal_1.complex != signal_2.complex:
+        raise ValueError(("Both signals must be complex or real valued. "
+                          "Mixtures are not allowed."))
+
+    if mode not in ['linear', 'cyclic']:
+        raise ValueError(f"mode is '{mode}' but must be 'linear' or 'cyclic'")
+
+    # copy input to avoid changing mutual data
+    signal_1 = signal_1.copy()
+    signal_2 = signal_2.copy()
```

Comment:
we dont need to copy here, since all functions below do their copies before processing. (e.g. by calling ```.flatten()```)